eng/common/scripts/artifact-metadata-parsing.ps1

```diff
@@ -101,11 +101,16 @@ function RetrieveReleaseTag($artifactLocation, $continueOnError = $true) {
   }
 }
 
-function RetrievePackages($artifactLocation) {
-  $pkgs = Get-ChildItem -Path $artifactLocation -Include $packagePattern -Recurse -File
+function RetrievePackages($artifactLocation, [switch]$NoRecurse) {
+  $pkgs = @()
+  if ($NoRecurse) {
+    $pkgs = Get-ChildItem -Path $artifactLocation -Include $packagePattern -Recurse -File -Depth 0
+  } else {
+    $pkgs = Get-ChildItem -Path $artifactLocation -Include $packagePattern -Recurse -File
+  }
   if ($GetPackageInfoFromPackageFileFn -and (Test-Path "Function:$GetPackageInfoFromPackageFileFn"))
   {
-    return $pkgs, $GetPackageInfoFromPackageFileFn
+    return @($pkgs), $GetPackageInfoFromPackageFileFn
```

Comment:
Can we not handle this inside of the language specific function when we call this?

Reply:
In the course of development I found a bug where if `$pkgs` is an empty array, powershell just ends up returning `$GetPackageInfoFromPackageFileFn` as a single value. So no matter what the caller is doing, the destructuring gets broken. This fixes that edge case issue (which was probably hidden since we have never failed to find a package before).

Reply:
So it's not a casting/type issue, but a return value issue.

Reply:
Sorry I wasn't referring to the `@(...)` part that I'm fine with I was more interested in the recurse option. We essentially call the language specific function with the set of packages can we not filter in there instead of not passing the recurse option?

Reply:
I don't think this is possible without some other refactoring that will still touch eng/common (happy to brainstorm, I'm not excited about this `NoRecurse` option anyway). The code paths here are:

```
- eng/common/scripts/create-tags-and-git-release.ps1
    - eng/common/scripts/artifact-metadata-parsing.ps1
        - <go repo>/eng/scripts/Language-Settings.ps1
```

When we do call into the language specific function, it's at https://github.com/Azure/azure-sdk-for-go/blob/main/eng/scripts/Language-Settings.ps1#L69-L72 for the `Get-<lang>-PackageInfoFromPackageFile` function, from https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/artifact-metadata-parsing.ps1#L125. The only relevant context given to this function, in go's case at least, is a `$pkg` object which is just a file type that points to the `go.mod` file. So we have no way, in this context, of doing logic to compare the go.mod path against the original service directory unless we plumb it through from the top level tag/git script like I've done for `NoRecurse`. It might be a better approach, but it has the downside of introducing optional parameters to how we call the language functions, and I'm not sure introducing more variability into the signature for an untyped, dynamically loaded function is a good idea.

Reply:
Do we ever care about this nested package? As in do we ever want to ship it? If not, could we fix this by changing the go.mod or marking it in some way as ignore?

Reply:
We do and it has its own pipeline as well, it's just the parent package/pipeline will also ship it if it's unreleased.

Reply:
I can see if a flattening of the directory structure would be possible, but my bet is quite a few things depend on it.